tests/testthat/test-renderer2-colour.R

```diff
@@ -0,0 +1,275 @@
+  data = mtcars,
+  alpha_off=0.5,
+  colour="red",
+  colour_off="transparent",
+  size = 10,
+  aes(x=wt, y=mpg,
+      fill = disp,
+      id=paste0("pointthree_disp", disp, "gear", gear, "wt", wt)),
+  clickSelects = "gear")+
+  ggtitle("colour_off + alpha_off"))
+
+info2 <- animint2HTML(viz.point)
+
+test_that("color_off only changes colour/stroke when clicked, fill does not change", {
+  point.xpath <- '//svg[@id="plot_pointthree"]//circle[@id="pointthree_disp275.8gear3wt3.73"]'
+  circle.list <- getNodeSet(info2$html, point.xpath)
+  before.click.color <- getStyleValue(info2$html, point.xpath, "stroke")
+  before.click.fill <- getStyleValue(info2$html, point.xpath, "fill")
+  
+  clickID('pointthree_disp275.8gear3wt3.73')
+  html <- getHTML()
+  after.click.color <- getStyleValue(html, point.xpath, "stroke")
+  after.click.fill <- getStyleValue(html, point.xpath, "fill")
+  
+  expect_false(isTRUE(all.equal(before.click.color, after.click.color)))
+  expect_equal(before.click.fill, after.click.fill)
+})
+
+test_that("fill and color are not same", {
+  point.xpath <- '//svg[@id="plot_pointthree"]//circle[@class="geom"]'
+  circle.list <- getNodeSet(info2$html, point.xpath)
+  circle.color <- getStyleValue(info2$html, point.xpath, "stroke")
+  circle.fill <- getStyleValue(info2$html, point.xpath, "fill")
+  expect_false(isTRUE(all.equal(circle.color, circle.fill)))
+})
+
+#
+# tests for g$geom="rect", originally only support 'stroke' as selection style
+#
+row.vec <- paste("row", c(1:3))
+col.vec <- paste("col", c(1:3))
+heat.data <- data.table(row.name = row.vec,
+  col.name = rep(col.vec, each=length(row.vec)),
+  value = c(2,8,-5,-7,15,3,-1,6,-7.5))
+no.col <- ggplot() +
+  geom_tile(data=heat.data,
+  aes(x = row.name, y = col.name, fill = value,
+  id=paste0("no_col_", value)),
+  clickSelects = "value")
+has.col.no.off <- ggplot() +
+  geom_tile(data=heat.data,
+  aes(x = row.name, y = col.name, fill = value,
+  id=paste0("col_", value)),
+  colour="red",
+  clickSelects = "value")
+has.col.and.off <- ggplot() +
+  geom_tile(data=heat.data,
+  aes(x = row.name, y = col.name, fill = value,
+  id=paste0("col_off_", value)),
+  colour="red",
+  colour_off="grey50",
+  clickSelects = "value")
+viz.tile <- list(
+  nocol=no.col,
+  colnooff=has.col.no.off,
+  colandoff=has.col.and.off)
+
+info <- animint2HTML(viz.tile)
+
+test_that("if has clickSelects but no colour/colour_off, selection colour/stroke should be black, and transparent for not selected (no stroke)", {
+  clickID('no_col_2')
+  html <- getHTML()
+  stroke.col <- getStyleValue(
+  info$html, '//g[@class="geom1_tile_nocol"]//rect[@id="no_col_2"]', "stroke")
+  expect.stroke.col <- "black"
+  expect_color(stroke.col, expect.stroke.col)
+  # not selected, stroke=transparent(no stroke style)
+  node.set <- getNodeSet(info$html, '//g[@class="geom1_tile_nocol"]//rect[@id="no_col_8"]')
+  style.strs <- as.character(sapply(node.set, function(x) xmlAttrs(x)["style"]))
```

Comment:
since you use this code block twice (once here and another time in the code below) can you please make a helper function? maybe something like
```r
expect_no_style(node.set, "stroke")
```

Reply:
yes!